src/Algebra/Graph/Relation/InternalDerived.hs

```diff
@@ -66,27 +59,146 @@ The 'Show' instance produces symmetrically closed expressions:
+-- 'Algebra.Graph.Relation.Symmetric.edgeCount'   (overlay x y) <= 'Algebra.Graph.Relation.Symmetric.edgeCount' x   + 'Algebra.Graph.Relation.Symmetric.edgeCount' y
+-- 'Algebra.Graph.Relation.Symmetric.vertexCount' (overlay 1 2) == 2
+-- 'Algebra.Graph.Relation.Symmetric.edgeCount'   (overlay 1 2) == 0
+-- @
+overlay :: Ord a => SymmetricRelation a -> SymmetricRelation a -> SymmetricRelation a
+overlay x y = SymmetricRelation $ Relation (domain (fromSymmetric x) `Set.union` domain (fromSymmetric y)) (relation (fromSymmetric x) `Set.union` relation (fromSymmetric y))
+
+-- | /Connect/ two graphs. This is an associative operation with the identity
+-- 'empty', which distributes over 'overlay' and obeys the decomposition axiom.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory. Note that the
+-- number of edges in the resulting graph is quadratic with respect to the number
+-- of vertices of the arguments: /m = O(m1 + m2 + n1 * n2 + n2 * n1)/.
+--
+-- @
+-- connect x y == connect y x
+-- 'Algebra.Graph.Relation.Symmetric.isEmpty'     (connect x y) == 'Algebra.Graph.Relation.Symmetric.isEmpty'   x   && 'Algebra.Graph.Relation.Symmetric.isEmpty'   y
+-- 'Algebra.Graph.Relation.Symmetric.hasVertex' z (connect x y) == 'Algebra.Graph.Relation.Symmetric.hasVertex' z x || 'Algebra.Graph.Relation.Symmetric.hasVertex' z y
+-- 'Algebra.Graph.Relation.Symmetric.vertexCount' (connect x y) >= 'Algebra.Graph.Relation.Symmetric.vertexCount' x
+-- 'Algebra.Graph.Relation.Symmetric.vertexCount' (connect x y) <= 'Algebra.Graph.Relation.Symmetric.vertexCount' x + 'Algebra.Graph.Relation.Symmetric.vertexCount' y
+-- 'Algebra.Graph.Relation.Symmetric.edgeCount'   (connect x y) >= 'Algebra.Graph.Relation.Symmetric.edgeCount' x
+-- 'Algebra.Graph.Relation.Symmetric.edgeCount'   (connect x y) >= 'Algebra.Graph.Relation.Symmetric.edgeCount' y
+-- 'Algebra.Graph.Relation.Symmetric.edgeCount'   (connect x y) >= 'Algebra.Graph.Relation.Symmetric.vertexCount' x * 'Algebra.Graph.Relation.Symmetric.vertexCount' y
+-- 'Algebra.Graph.Relation.Symmetric.vertexCount' (connect 1 2) == 2
+-- 'Algebra.Graph.Relation.Symmetric.edgeCount'   (connect 1 2) == 2
+-- @
+connect :: Ord a => SymmetricRelation a -> SymmetricRelation a -> SymmetricRelation a
+connect x y = SymmetricRelation $ Relation (domain (fromSymmetric x) `Set.union` domain (fromSymmetric y))
+    (relation (fromSymmetric x) `Set.union` relation (fromSymmetric y) `Set.union` (domain (fromSymmetric x) `setProduct` domain (fromSymmetric y)) `Set.union` (domain (fromSymmetric y) `setProduct` domain (fromSymmetric x)))
+
+instance (Ord a, Num a) => Num (SymmetricRelation a) where
+    fromInteger = vertex . fromInteger
+    x + y       = x `overlay` y
+    x * y       = x `connect` y
+    signum      = const empty
+    abs         = id
+    negate      = id
+
+-- | Check if the internal representation of a relation is consistent, i.e. if all
```

Comment:
I think it may be best to store both directions of edges in the `Relation`, in which case this invariant should be included into the definition of `consistent`. By forcing to store edges in both directions, you will get rid of a lot of calls to `fromSymmetric` in this module, which is likely going to be beneficial performance wise for common use-cases.

Reply:
I actually do that, because the basic primitives, like `connect` or `(*)` do that, so `fromSymmetric` is O(1), right?

Reply:
What I should have removed was the `symmetricClosure` call in the `Eq` and `Show` instances because it's no longer needed.

Reply:
I see, it was my misunderstanding of how `connect` works. Thanks for clarifying!

Reply:
Since you do this already, you should definitely include this new invariant into the definition of `consistent`.

Reply:
How should I do that? Are you referring to the documentation?

Reply:
I'm referring to both documentation and the code. 

The code should test that you cannot violate the invariant that you always have both directions of an edge stored. One way to test this, I think, is to say that `x == transpose x`.

Reply:
Got it!